File: android/src/main/java/com/scandit/datacapture/flutter/barcode/tracking/BarcodeTrackingMethodHandler.java
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2024- Scandit AG. All rights reserved.
 */
package com.scandit.datacapture.flutter.barcode.tracking;

import androidx.annotation.NonNull;

import com.scandit.datacapture.flutter.core.utils.FlutterResult;
import com.scandit.datacapture.flutter.core.utils.ResultUtils;
import com.scandit.datacapture.frameworks.barcode.tracking.BarcodeTrackingModule;
import com.scandit.datacapture.frameworks.core.FrameworkModule;
import com.scandit.datacapture.frameworks.core.errors.FrameDataNullError;
import com.scandit.datacapture.frameworks.core.locator.ServiceLocator;
import com.scandit.datacapture.frameworks.core.utils.DefaultLastFrameData;
import com.scandit.datacapture.frameworks.core.utils.LastFrameData;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

import org.json.JSONObject;

import java.util.HashMap;

public class BarcodeTrackingMethodHandler implements MethodChannel.MethodCallHandler {

    public static final String EVENT_CHANNEL_NAME = "com.scandit.datacapture.barcode.tracking/event_channel";
    public static final String METHOD_CHANNEL_NAME = "com.scandit.datacapture.barcode.tracking/method_channel";

    private final ServiceLocator<FrameworkModule> serviceLocator;
    private final LastFrameData lastFrameData;

    public BarcodeTrackingMethodHandler(ServiceLocator<FrameworkModule> serviceLocator) {
        this(serviceLocator, DefaultLastFrameData.getInstance());
    }

    public BarcodeTrackingMethodHandler(ServiceLocator<FrameworkModule> serviceLocator, LastFrameData lastFrameData) {
        this.serviceLocator = serviceLocator;
        this.lastFrameData = lastFrameData;
    }

    @Override
    public void onMethodCall(MethodCall call, @NonNull MethodChannel.Result result) {
        switch (call.method) {
            case "getBarcodeTrackingDefaults":
                result.success(new JSONObject(getSharedModule().getDefaults()).toString());
                break;
            case "addBarcodeTrackingListener":
                getSharedModule().addBarcodeTrackingListener();
                result.success(null);
                break;
            case "removeBarcodeTrackingListener":
                getSharedModule().removeBarcodeTrackingListener();
                result.success(null);
                break;
            case "barcodeTrackingFinishDidUpdateSession":
                getSharedModule().finishDidUpdateSession(Boolean.TRUE.equals(call.arguments));
                result.success(true);
                break;
            case "resetBarcodeTrackingSession":
                getSharedModule().resetSession(
                        call.arguments instanceof Long ? (Long) call.arguments : null
                );
                result.success(null);
                break;
            case "getLastFrameData":
                lastFrameData.getLastFrameDataBytes((bytes) -> {
                    if (bytes == null) {
                        ResultUtils.rejectKotlinError(result, new FrameDataNullError());
                        return null;
                    }
                    result.success(bytes);
                    return null;
                });
                break;
            case "addBarcodeTrackingAdvancedOverlayDelegate":
                getSharedModule().addAdvancedOverlayListener();
                result.success(null);
                break;
            case "removeBarcodeTrackingAdvancedOverlayDelegate":
                getSharedModule().removeAdvancedOverlayListener();
                result.success(null);
                break;
            case "clearTrackedBarcodeWidgets":
                getSharedModule().clearAdvancedOverlayTrackedBarcodeViews();
                result.success(null);
                break;
            case "setWidgetForTrackedBarcode":
                if (!(call.arguments instanceof HashMap)) {
                    result.error("-1", "Invalid argument for setWidgetForTrackedBarcode", "");
                    return;
                }

                //noinspection unchecked
                getSharedModule().setWidgetForTrackedBarcode((HashMap<String, Object>) call.arguments);
                result.success(null);
                break;
            case "setAnchorForTrackedBarcode":
                if (!(call.arguments instanceof HashMap)) {
                    result.error("-1", "Invalid argument for setAnchorForTrackedBarcode", "");
                    return;
                }
                //noinspection unchecked
                getSharedModule().setAnchorForTrackedBarcode((HashMap<String, Object>) call.arguments);
                result.success(null);
                break;
            case "setOffsetForTrackedBarcode":
                if (!(call.arguments instanceof HashMap)) {
                    result.error("-1", "Invalid argument for setOffsetForTrackedBarcode", "");
                    return;
                }
                //noinspection unchecked
                getSharedModule().setOffsetForTrackedBarcode((HashMap<String, Object>) call.arguments);
                result.success(null);
                break;
            case "subscribeBarcodeTrackingBasicOverlayListener":
                getSharedModule().addBasicOverlayListener();
                result.success(null);
                break;
            case "unsubscribeBarcodeTrackingBasicOverlayListener":
                getSharedModule().removeBasicOverlayListener();
                result.success(null);
                break;
            case "setBrushForTrackedBarcode":
                getSharedModule().setBasicOverlayBrushForTrackedBarcode(
                        call.arguments.toString()
                );
                result.success(null);
                break;
            case "clearTrackedBarcodeBrushes":
                getSharedModule().clearBasicOverlayTrackedBarcodeBrushes();
                result.success(null);
                break;
            case "setModeEnabledState":
                getSharedModule().setModeEnabled(Boolean.TRUE.equals(call.arguments()));
                break;
            case "updateBarcodeTrackingMode":
                assert call.arguments() != null;
                getSharedModule().updateModeFromJson(
                        call.arguments(),
                        new FlutterResult(result)
                );
                break;
            case "applyBarcodeTrackingModeSettings":
                assert call.arguments() != null;
                getSharedModule().applyModeSettings(
                        call.arguments(),
                        new FlutterResult(result)
                );
                break;
            case "updateBarcodeTrackingBasicOverlay":
                assert call.arguments() != null;
                getSharedModule().updateBasicOverlay(
                        call.arguments(),
                        new FlutterResult(result)
                );
                break;
            case "updateBarcodeTrackingAdvancedOverlay":
                assert call.arguments() != null;
                getSharedModule().updateAdvancedOverlay(
                        call.arguments(),
                        new FlutterResult(result)
                );
                break;
            default:
                result.notImplemented();
        }
    }

    private volatile BarcodeTrackingModule sharedModuleInstance;

    private BarcodeTrackingModule getSharedModule() {
        if (sharedModuleInstance == null) {
            synchronized (this) {
                if (sharedModuleInstance == null) {
                    sharedModuleInstance = (BarcodeTrackingModule)this.serviceLocator.resolve(BarcodeTrackingModule.class.getName());
                }
            }
        }
        return sharedModuleInstance;
    }
}

File: android/src/main/kotlin/com/scandit/datacapture/flutter/barcode/tracking/BarcodeTrackingMethodHandler.kt
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2020- Scandit AG. All rights reserved.
 */
package com.scandit.datacapture.flutter.barcode.tracking

import com.scandit.datacapture.flutter.core.utils.FlutterResult
import com.scandit.datacapture.flutter.core.utils.rejectKotlinError
import com.scandit.datacapture.frameworks.barcode.tracking.BarcodeTrackingModule
import com.scandit.datacapture.frameworks.core.errors.FrameDataNullError
import com.scandit.datacapture.frameworks.core.utils.DefaultLastFrameData
import com.scandit.datacapture.frameworks.core.utils.LastFrameData
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

/** ScanditFlutterDataCaptureBarcodeTrackingPlugin */
class BarcodeTrackingMethodHandler(
    private val barcodeTrackingModule: BarcodeTrackingModule,
    private val lastFrameData: LastFrameData = DefaultLastFrameData.getInstance()
) : MethodChannel.MethodCallHandler {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            METHOD_GET_BARCODE_TRACKING_DEFAULTS ->
                result.success(JSONObject(barcodeTrackingModule.getDefaults()).toString())
            METHOD_ADD_BARCODE_TRACKING_LISTENER -> {
                barcodeTrackingModule.addBarcodeTrackingListener()
                result.success(null)
            }
            METHOD_REMOVE_BARCODE_TRACKING_LISTENER -> {
                barcodeTrackingModule.removeBarcodeTrackingListener()
                result.success(null)
            }
            METHOD_TRACKING_FINISH_DID_UPDATE_SESSION -> {
                barcodeTrackingModule.finishDidUpdateSession(
                    call.arguments as? Boolean ?: false
                )
                result.success(true)
            }
            METHOD_RESET_TRACKING_SESSION -> {
                barcodeTrackingModule.resetSession(call.arguments as? Long)
                result.success(null)
            }
            METHOD_GET_LAST_FRAME_DATA -> lastFrameData.getLastFrameDataBytes {
                if (it == null) {
                    result.rejectKotlinError(FrameDataNullError())
                    return@getLastFrameDataBytes
                }
                result.success(it)
            }
            METHOD_ADD_BARCODE_TRACKING_ADVANCED_OVERLAY_LISTENER -> {
                barcodeTrackingModule.addAdvancedOverlayListener()
                result.success(null)
            }
            METHOD_REMOVE_BARCODE_TRACKING_ADVANCED_OVERLAY_LISTENER -> {
                barcodeTrackingModule.removeAdvancedOverlayListener()
                result.success(null)
            }
            METHOD_CLEAR_TRACKED_BARCODE_WIDGETS -> {
                barcodeTrackingModule.clearAdvancedOverlayTrackedBarcodeViews()
                result.success(null)
            }
            METHOD_SET_WIDGET_FOR_TRACKED_BARCODE -> {
                @Suppress("UNCHECKED_CAST")
                val arguments = call.arguments as? HashMap<String, Any?> ?: run {
                    result.error(
                        "-1",
                        "Invalid argument for $METHOD_SET_WIDGET_FOR_TRACKED_BARCODE",
                        ""
                    )
                    return
                }
                barcodeTrackingModule.setWidgetForTrackedBarcode(arguments)
                result.success(null)
            }
            METHOD_SET_ANCHOR_FOR_TRACKED_BARCODE -> {
                @Suppress("UNCHECKED_CAST")
                val arguments = call.arguments as? HashMap<String, Any?> ?: run {
                    result.error(
                        "-1",
                        "Invalid argument for $METHOD_SET_ANCHOR_FOR_TRACKED_BARCODE",
                        ""
                    )
                    return
                }
                barcodeTrackingModule.setAnchorForTrackedBarcode(arguments)
                result.success(null)
            }
            METHOD_SET_OFFSET_FOR_TRACKED_BARCODE -> {
                @Suppress("UNCHECKED_CAST")
                val arguments = call.arguments as? HashMap<String, Any?> ?: run {
                    result.error(
                        "-1",
                        "Invalid argument for $METHOD_SET_OFFSET_FOR_TRACKED_BARCODE",
                        ""
                    )
                    return
                }
                barcodeTrackingModule.setOffsetForTrackedBarcode(arguments)
                result.success(null)
            }
            METHOD_SUBSCRIBE_BASIC_OVERLAY_LISTENER -> {
                barcodeTrackingModule.addBasicOverlayListener()
                result.success(null)
            }
            METHOD_UNSUBSCRIBE_BASIC_OVERLAY_LISTENER -> {
                barcodeTrackingModule.removeBasicOverlayListener()
                result.success(null)
            }
            METHOD_SET_BRUSH_FOR_TRACKED_BARCODE -> {
                barcodeTrackingModule.setBasicOverlayBrushForTrackedBarcode(
                    call.arguments as String
                )
                result.success(null)
            }
            METHOD_CLEAR_TRACKED_BARCODE_BRUSHES -> {
                barcodeTrackingModule.clearBasicOverlayTrackedBarcodeBrushes()
                result.success(null)
            }
            METHOD_SET_MODE_ENABLED_STATE -> barcodeTrackingModule.setModeEnabled(
                call.arguments as Boolean
            )
            METHOD_UPDATE_MODE_FROM_JSON -> barcodeTrackingModule.updateModeFromJson(
                call.arguments as String,
                FlutterResult(result)
            )
            METHOD_APPLY_MODE_SETTINGS -> barcodeTrackingModule.applyModeSettings(
                call.arguments as String,
                FlutterResult(result)
            )
            METHOD_UPDATE_BASIC_OVERLAY -> barcodeTrackingModule.updateBasicOverlay(
                call.arguments as String,
                FlutterResult(result)
            )
            METHOD_UPDATE_ADVANCED_OVERLAY -> barcodeTrackingModule.updateAdvancedOverlay(
                call.arguments as String,
                FlutterResult(result)
            )
            else -> result.notImplemented()
        }
    }

    companion object {
        private const val METHOD_GET_BARCODE_TRACKING_DEFAULTS = "getBarcodeTrackingDefaults"
        private const val METHOD_ADD_BARCODE_TRACKING_LISTENER = "addBarcodeTrackingListener"
        private const val METHOD_REMOVE_BARCODE_TRACKING_LISTENER = "removeBarcodeTrackingListener"
        private const val METHOD_TRACKING_FINISH_DID_UPDATE_SESSION =
            "barcodeTrackingFinishDidUpdateSession"
        private const val METHOD_RESET_TRACKING_SESSION = "resetBarcodeTrackingSession"
        private const val METHOD_GET_LAST_FRAME_DATA = "getLastFrameData"
        private const val METHOD_SET_WIDGET_FOR_TRACKED_BARCODE = "setWidgetForTrackedBarcode"
        private const val METHOD_SET_ANCHOR_FOR_TRACKED_BARCODE = "setAnchorForTrackedBarcode"
        private const val METHOD_SET_OFFSET_FOR_TRACKED_BARCODE = "setOffsetForTrackedBarcode"
        private const val METHOD_CLEAR_TRACKED_BARCODE_WIDGETS = "clearTrackedBarcodeWidgets"

        private const val METHOD_ADD_BARCODE_TRACKING_ADVANCED_OVERLAY_LISTENER =
            "addBarcodeTrackingAdvancedOverlayDelegate"
        private const val METHOD_REMOVE_BARCODE_TRACKING_ADVANCED_OVERLAY_LISTENER =
            "removeBarcodeTrackingAdvancedOverlayDelegate"
        private const val METHOD_SET_BRUSH_FOR_TRACKED_BARCODE = "setBrushForTrackedBarcode"
        private const val METHOD_CLEAR_TRACKED_BARCODE_BRUSHES = "clearTrackedBarcodeBrushes"
        private const val METHOD_SUBSCRIBE_BASIC_OVERLAY_LISTENER =
            "subscribeBarcodeTrackingBasicOverlayListener"
        private const val METHOD_UNSUBSCRIBE_BASIC_OVERLAY_LISTENER =
            "unsubscribeBarcodeTrackingBasicOverlayListener"
        private const val METHOD_SET_MODE_ENABLED_STATE = "setModeEnabledState"
        private const val METHOD_UPDATE_MODE_FROM_JSON = "updateBarcodeTrackingMode"
        private const val METHOD_APPLY_MODE_SETTINGS = "applyBarcodeTrackingModeSettings"
        private const val METHOD_UPDATE_BASIC_OVERLAY = "updateBarcodeTrackingBasicOverlay"
        private const val METHOD_UPDATE_ADVANCED_OVERLAY = "updateBarcodeTrackingAdvancedOverlay"

        const val EVENT_CHANNEL_NAME =
            "com.scandit.datacapture.barcode.tracking/event_channel"

        const val METHOD_CHANNEL_NAME =
            "com.scandit.datacapture.barcode.tracking/method_channel"
    }
}
